app/scripts/controllers/transactions/index.js

```diff
@@ -331,6 +331,195 @@ export default class TransactionController extends EventEmitter {
+   * @param txGasFees.maxPriorityFeePerGas
+   * @param txGasFees.maxFeePerGas
+   * @param txGasFees.estimateUsed
+   */
+  updateTransactionGasFees(
+    txId,
+    { gasLimit, gasPrice, maxPriorityFeePerGas, maxFeePerGas, estimateUsed },
+  ) {
+    if (!this._validateTransaction(txId)) {
+      return;
+    }
+
+    const txGasFees = {
+      txParams: {
+        gasLimit,
+        gasPrice,
+        maxPriorityFeePerGas,
+        maxFeePerGas,
+      },
+      estimateUsed,
+    };
+
+    const note = `Update Transaction Gas Fees for ${txId}`;
+    this._updateTransaction(txId, txGasFees, note);
+  }
+
+  /**
+   * updates the estimate base fees of the transaction with id if the transaction state is unapproved
+   *
+   * @param {string} txId - transaction id
+   * @param {object} txEstimateBaseFees - holds the estimate base fees parameters
+   * {
+   * estimatedBaseFee,
+   * decEstimatedBaseFee
+   * }
+   * @param txEstimateBaseFees.estimatedBaseFee
+   * @param txEstimateBaseFees.decEstimatedBaseFee
+   */
+  updateTransactionEstimateBaseFee(
+    txId,
+    { estimatedBaseFee, decEstimatedBaseFee },
+  ) {
+    if (!this._validateTransaction(txId)) {
+      return;
+    }
+
+    const txEstimateBaseFees = {
+      estimatedBaseFee,
+      decEstimatedBaseFee,
+    };
+
+    const note = `Update Transaction Estimated Base Fees for ${txId}`;
+    this._updateTransaction(txId, txEstimateBaseFees, note);
+  }
+
+  /**
+   * updates a swap approval transaction with provided metadata and source token symbol
+   *  if the transaction state is unapproved.
+   *
+   * @param {string} txId
+   * @param {object} swapApprovalTransaction - holds the metadata and token symbol
+   * {
+   * type,
+   * sourceTokenSymbol
+   * }
+   * @param swapApprovalTransaction.type
+   * @param swapApprovalTransaction.sourceTokenSymbol
+   */
+  updateSwapApprovalTransaction(txId, { type, sourceTokenSymbol }) {
+    if (!this._validateTransaction(txId)) {
+      return;
+    }
+
+    const swapApprovalTransaction = { type, sourceTokenSymbol };
+    const note = `Update Swap Approval Transaction for ${txId}`;
+    this._updateTransaction(txId, swapApprovalTransaction, note);
+  }
+
+  /**
+   * updates a swap transaction with provided metadata and source token symbol
+   *  if the transaction state is unapproved.
+   *
+   * @param {string} txId
+   * @param {object} swapTransaction - holds the metadata
+   * {
+   * sourceTokenSymbol,
+   * destinationTokenSymbol,
+   * type,
+   * destinationTokenDecimals,
+   * destinationTokenAddress,
+   * swapMetaData,
+   * swapTokenValue,
+   *}
+   * @param swapTransaction.sourceTokenSymbol
+   * @param swapTransaction.destinationTokenSymbol
+   * @param swapTransaction.type
+   * @param swapTransaction.destinationTokenDecimals
+   * @param swapTransaction.destinationTokenAddress
+   * @param swapTransaction.swapMetaData
+   * @param swapTransaction.swapTokenValue
+   */
+  updateSwapTransaction(
```

Comment:
We should make a swap transaction in the app and then get the state logs to verify the shape of the state logs

Reply:
This is Swap Data

```
{
    "id": 5280220671892615,
    "time": 1644423917210,
    "status": "confirmed",
    "metamaskNetworkId": "4",
    "originalGasEstimate": "0x9810",
    "userEditedGasLimit": false,
    "chainId": "0x4",
    "loadingDefaults": false,
    "dappSuggestedGasFees": null,
    "txParams": {
        "from": "0xdec94d958b0da848d8cbfa0a9015451a4625452e",
        "to": "0x881d40237659c251811cec9c364ef91dc08d300c",
        "nonce": "0x0",
        "value": "0xde0b6b3a7640000",
        "gas": "0x9810",
        "maxFeePerGas": "0x178ee1aa8",
        "maxPriorityFeePerGas": "0x77359400"
    },
    "origin": "metamask",
    "type": "swap",
    "userFeeLevel": "medium",
    "defaultGasEstimates": {
        "estimateType": "medium",
        "gas": "0x9810",
        "maxFeePerGas": "0x178ee1aa8",
        "maxPriorityFeePerGas": "0x77359400"
    },
    "estimatedBaseFee": "c63f2c81",
    "sourceTokenSymbol": "ETH",
    "destinationTokenSymbol": "LUNA",
    "destinationTokenDecimals": 18,
    "destinationTokenAddress": "0xd2877702675e6ceb975b4a1dff9fb7baf4c91ea9",
    "swapMetaData": {
        "token_from": "ETH",
        "token_from_amount": "1",
        "token_to": "LUNA",
        "token_to_amount": "55.540352",
        "slippage": 3,
        "custom_slippage": true,
        "best_quote_source": "zeroEx",
        "other_quote_selected": false,
        "other_quote_selected_source": "",
        "gas_fees": "0.550904",
        "estimated_gas": "32440",
        "used_gas_price": "0",
        "is_hardware_wallet": false,
        "max_fee_per_gas": "178ee1aa8",
        "max_priority_fee_per_gas": "77359400",
        "base_and_priority_fee_per_gas": "13d74c081"
    },
    "swapTokenValue": "1",
    "preTxBalance": "29a2241af62c0000",
    "hash": "0x51e4270966c41b8ef39d227ae53902ab0c5574e2d57b383e1ad7bc32a30a5949",
    "submittedTime": 1644423919031,
    "firstRetryBlockNumber": "0x9ab9be",
    "baseFeePerGas": "0xb3e84c6a",
    "blockTimestamp": "6203eaf3",
    "postTxBalance": "1bc0d9579cda87d0"
}
```

Reply:
@danjm I see that the method takes a swapMetaData 
```
    "swapMetaData": {
        "token_from": "ETH",
        "token_from_amount": "1",
        "token_to": "LUNA",
        "token_to_amount": "55.540352",
        "slippage": 3,
        "custom_slippage": true,
        "best_quote_source": "zeroEx",
        "other_quote_selected": false,
        "other_quote_selected_source": "",
        "gas_fees": "0.550904",
        "estimated_gas": "32440",
        "used_gas_price": "0",
        "is_hardware_wallet": false,
        "max_fee_per_gas": "178ee1aa8",
        "max_priority_fee_per_gas": "77359400",
        "base_and_priority_fee_per_gas": "13d74c081"
    },
 ```
 
 Do we need to validate this too?

Reply:
I don't think we need to validate that

Reply:
Thanks for sharing the transaction metadata. Based on that, it looks like the structure of the data here is correct